[Enhancement] Improve ORC column reader's performance and add benchmark

Smith-Cruise opened the pull request:
## Doing List
* Use deflate to decompression zlib in orc
* Add benchmark for orc column reader
* Optimize orc column reader performance

## Original Performance:
```bash
----------------------------------------------------------------------------------------------------------------
Benchmark                                                                      Time             CPU   Iterations
----------------------------------------------------------------------------------------------------------------
BM_primitive<NULLABLE, LogicalType::TYPE_BOOLEAN>/iterations:5              26.6 ms         26.6 ms            5
BM_primitive<NON_NULLABLE, LogicalType::TYPE_BOOLEAN>/iterations:5          19.3 ms         19.3 ms            5
BM_primitive<NULLABLE, LogicalType::TYPE_VARCHAR>/iterations:5              90.6 ms         90.6 ms            5
BM_primitive<NON_NULLABLE, LogicalType::TYPE_VARCHAR>/iterations:5           221 ms          221 ms            5
BM_primitive<NULLABLE, LogicalType::TYPE_CHAR>/iterations:5                 90.6 ms         90.6 ms            5
BM_primitive<NON_NULLABLE, LogicalType::TYPE_CHAR>/iterations:5              437 ms          437 ms            5
BM_primitive<NULLABLE, LogicalType::TYPE_INT>/iterations:5                  23.7 ms         23.7 ms            5
BM_primitive<NON_NULLABLE, LogicalType::TYPE_INT>/iterations:5              11.7 ms         11.7 ms            5
BM_primitive<NULLABLE, LogicalType::TYPE_DOUBLE>/iterations:5               31.3 ms         31.3 ms            5
BM_primitive<NON_NULLABLE, LogicalType::TYPE_DOUBLE>/iterations:5           76.1 ms         76.1 ms            5
BM_primitive<NULLABLE, LogicalType::TYPE_DECIMAL64>/iterations:5            38.1 ms         38.1 ms            5
BM_primitive<NON_NULLABLE, LogicalType::TYPE_DECIMAL64>/iterations:5        91.6 ms         91.6 ms            5
BM_primitive<NULLABLE, LogicalType::TYPE_DECIMAL128>/iterations:5           48.6 ms         48.6 ms            5
BM_primitive<NON_NULLABLE, LogicalType::TYPE_DECIMAL128>/iterations:5        493 ms          493 ms            5
BM_primitive<NULLABLE, LogicalType::TYPE_VARBINARY>/iterations:5            58.2 ms         58.2 ms            5
BM_primitive<NON_NULLABLE, LogicalType::TYPE_VARBINARY>/iterations:5         225 ms          225 ms            5
BM_primitive<NULLABLE, LogicalType::TYPE_DATE>/iterations:5                 31.0 ms         31.0 ms            5
BM_primitive<NON_NULLABLE, LogicalType::TYPE_DATE>/iterations:5             14.5 ms         14.5 ms            5
BM_primitive<NULLABLE, LogicalType::TYPE_DATETIME>/iterations:5             47.6 ms         47.6 ms            5
BM_primitive<NON_NULLABLE, LogicalType::TYPE_DATETIME>/iterations:5          102 ms          102 ms            5
```

## After Optimize
```bash
---------------------------------------------------------------------------------------------------------------- 
Benchmark                                                                      Time             CPU   Iterations 
---------------------------------------------------------------------------------------------------------------- 
BM_primitive<NULLABLE, LogicalType::TYPE_BOOLEAN>/iterations:5              23.0 ms         23.0 ms            5 115.65%
BM_primitive<NON_NULLABLE, LogicalType::TYPE_BOOLEAN>/iterations:5          19.1 ms         19.1 ms            5 101.05%
BM_primitive<NULLABLE, LogicalType::TYPE_VARCHAR>/iterations:5              45.5 ms         45.5 ms            5 199.12%
BM_primitive<NON_NULLABLE, LogicalType::TYPE_VARCHAR>/iterations:5           113 ms          113 ms            5 195.58%
BM_primitive<NULLABLE, LogicalType::TYPE_CHAR>/iterations:5                 43.8 ms         43.8 ms            5 206.85%
BM_primitive<NON_NULLABLE, LogicalType::TYPE_CHAR>/iterations:5              306 ms          306 ms            5 142.81%
BM_primitive<NULLABLE, LogicalType::TYPE_INT>/iterations:5                  20.4 ms         20.4 ms            5 116.18%
BM_primitive<NON_NULLABLE, LogicalType::TYPE_INT>/iterations:5              10.7 ms         10.7 ms            5 109.35%
BM_primitive<NULLABLE, LogicalType::TYPE_DOUBLE>/iterations:5               24.0 ms         24.0 ms            5 130.42%
BM_primitive<NON_NULLABLE, LogicalType::TYPE_DOUBLE>/iterations:5           75.3 ms         75.3 ms            5 101.06%
BM_primitive<NULLABLE, LogicalType::TYPE_DECIMAL64>/iterations:5            26.4 ms         26.4 ms            5 144.32%
BM_primitive<NON_NULLABLE, LogicalType::TYPE_DECIMAL64>/iterations:5        86.7 ms         86.7 ms            5 105.65%
BM_primitive<NULLABLE, LogicalType::TYPE_DECIMAL128>/iterations:5           37.9 ms         37.9 ms            5 128.23%
BM_primitive<NON_NULLABLE, LogicalType::TYPE_DECIMAL128>/iterations:5        491 ms          491 ms            5 100.41%
BM_primitive<NULLABLE, LogicalType::TYPE_VARBINARY>/iterations:5            45.0 ms         45.0 ms            5 129.33%
BM_primitive<NON_NULLABLE, LogicalType::TYPE_VARBINARY>/iterations:5         126 ms          126 ms            5 178.57%
BM_primitive<NULLABLE, LogicalType::TYPE_DATE>/iterations:5                 25.7 ms         25.7 ms            5 120.62%
BM_primitive<NON_NULLABLE, LogicalType::TYPE_DATE>/iterations:5             8.60 ms         8.60 ms            5 168.60%
BM_primitive<NULLABLE, LogicalType::TYPE_DATETIME>/iterations:5             39.3 ms         39.3 ms            5 121.12%
BM_primitive<NON_NULLABLE, LogicalType::TYPE_DATETIME>/iterations:5         95.5 ms         95.5 ms            5 106.81%
```

## TPCH
100g tpch zlib orc : select * from lineitem order by limit 1

dop=2, io tasks per scan operator=2

**Origin performance**
65s

**After improve**
47s

## What type of PR is this:
- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Checklist:
- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5
  - [ ] 2.4

wanpengfei-git commented:
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

wanpengfei-git commented:
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

wanpengfei-git commented:
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

wanpengfei-git commented:
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

wanpengfei-git commented:
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

wanpengfei-git commented:
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

wanpengfei-git commented:
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

wanpengfei-git commented:
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

letian-jiang commented on be/src/bench/orc_column_reader_bench.cpp:
```diff
@@ -0,0 +1,357 @@
+
+// Run command: ./be/build_Release/src/bench/output/orc_column_reader_bench 2>>/dev/null
+
+const static size_t benchmarkIterationTimes = 5;
```
5 is relatively small.

Maybe we can use `MinTime()` to measure iterations of each setting.

>   // Set the minimum amount of time to use when running this benchmark. This
  // option overrides the `benchmark_min_time` flag.
  // REQUIRES: `t > 0` and `Iterations` has not been called on this benchmark.
  Benchmark* MinTime(double t);

Smith-Cruise replied:
If we set larger iteration time, the later time-consuming will be shorter and shorter because of the cpu cache

letian-jiang commented on be/src/formats/orc/apache-orc/c++/src/Compression.cc:
```diff
@@ -723,6 +728,38 @@ void BlockDecompressionStream::NextDecompress(const void** data, int* size, size
+    LibDeflateDecompressionStream(std::unique_ptr<SeekableInputStream> inStream, size_t blockSize, MemoryPool& _pool,
+                                  ReaderMetrics* _metrics)
+            : BlockDecompressionStream(std::move(inStream), blockSize, _pool, _metrics) {
+        decompressor = libdeflate_alloc_decompressor();
```
check alloc failure here?

Smith-Cruise replied:
added

DorianZheng commented on be/src/formats/orc/column_reader.cpp:
```diff
@@ -592,42 +533,60 @@ Status StringColumnReader::get_next(orc::ColumnVectorBatch* cvb, ColumnPtr& col,
+                // set the correct size for vb
+                vb.resize(write_pos);
             } else {
+                size_t write_pos = vb.size();
```
```
memcpy(vb.data, data->data, data->size);
```

Smith-Cruise replied:
orc's StringVectorBatch structure:
```c++
// pointers to the start of each string
DataBuffer<char*> data;
// the length of each string
DataBuffer<int64_t> length;
// string blob
DataBuffer<char> blob;
// dict codes, iff. there is dictionary.
DataBuffer<int64_t> codes;
```
We can't make sure each line's string is continuous in `blob`.

DorianZheng commented on be/src/formats/orc/column_reader.cpp:
```diff
@@ -747,34 +704,34 @@ Status VarbinaryColumnReader::get_next(orc::ColumnVectorBatch* cvb, ColumnPtr& c
-            for (int i = col_start; i < col_start + size; ++i, ++pos) {
-                vb.insert(vb.end(), data->data[pos], data->data[pos] + data->length[pos]);
-                vo.emplace_back(vb.size());
+            size_t write_pos = vb.size();
```
ditto

Smith-Cruise committed:
[Enhancement] Improve orc column reader performance and add benchmark

Signed-off-by: Smith Cruise <chendingchao1@126.com>
tmp

Signed-off-by: Smith Cruise <chendingchao1@126.com>

tmp

Signed-off-by: Smith Cruise <chendingchao1@126.com>

tmp

Signed-off-by: Smith Cruise <chendingchao1@126.com>

tmp

Signed-off-by: Smith Cruise <chendingchao1@126.com>

tmp

Signed-off-by: Smith Cruise <chendingchao1@126.com>

tmp

Signed-off-by: Smith Cruise <chendingchao1@126.com>

improve string

Signed-off-by: Smith Cruise <chendingchao1@126.com>

improve varbinary

Signed-off-by: Smith Cruise <chendingchao1@126.com>

improve date

Signed-off-by: Smith Cruise <chendingchao1@126.com>

improve date

Signed-off-by: Smith Cruise <chendingchao1@126.com>

dirtysalt commented on be/src/formats/orc/apache-orc/c++/src/Compression.cc:
```diff
@@ -1110,8 +1150,18 @@ std::unique_ptr<SeekableInputStream> createDecompressor(CompressionKind kind,
         return REDUNDANT_MOVE(input);
     case CompressionKind_ZLIB:
+#ifdef __x86_64__
+        if (starrocks::config::enable_orc_libdeflate_decompression) {
```
Have you benchmarked libdeflate is better?

wanpengfei-git commented:
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)
